A36: xDS-Enabled Servers

ejona86 opened the pull request:
See "open issues" at the end of the doc. At the time of creation, Go and C are much more tentative than Java.

CC @sanjaypujare, @dfawley, @markdroth

ejona86 commented:
CC @easwars

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+
+The xDS-returned Listener must have an [`address`][Listener.address] that
+matches the listening address provided. The Listener's `address` would be a TCP
+`SocketAddress` with matching `address` and `port_value`. The XdsClient must NAK
```
s/NAK/NACK maybe?

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+`example/resource?udpa.resource.listening_address=[::]:80`. Note that the `[]`
+is _not_ percent encoded, as this string is not a URI.
+
+The xDS-returned Listener must have an [`address`][Listener.address] that
```
In Go, the piece of code which processes the response from the management server and decides whether to ack or nack, does not know about the set of resources being watched (that happens at a higher layer). This means that when we get a `Listener` resource, we do not have a way to make sure that the `address` and `port_value` in the `SocketAddress` match the one specified in the `listening_address` field. What we *can* do is to make sure that it matches with the `name` field (because eventually the `name` in the response has to match the one in the request).

I wanted to bring this us up to see if there will be any inconsistency in the handling of this across the languages.

sanjaypujare replied:
There is no `listening_address` field. It is just encoded in the resource name in the request. So when you match the name in the response to the name in the request you are indirectly verifying the value of the `listening_address`. In that case it should not be difficult to match that part with the `[address][Listener.address]` value either.

markdroth replied:
Repeating my earlier comment from reviewable:
> Hmm, good question. Maybe the `XdsClient` can verify that the `Listener` resource has its address field set to the same address as in the resource name. Although that might be difficult if we use the template approach I suggest above instead of hard-coding "xds.resource.listener_address=" in our implementation.

We have now switched to the template approach.  Given that, I don't think we can make any assumptions about the `name` field.

I would like to do this validation, but I don't see a reasonable way to do it without some fairly ugly layering violations.  Unless one of you can think of another way to do this, I think we may need to live without this check. :(

sanjaypujare replied:
I can think of the following layered approach for validating this:

- the xDSClient validates the returned Listener by matching the resource name against the requested resource name (trivial common check for all resources?)
- the actual watcher that requested this Listener resource then validates that the returned Listener has the [`address`][Listener.address] matching the expected value. This might be too late to NACK to the control plane but should also be rare (a bug in the control plane).

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+
+### Go
+
+Create an `xds.Server` struct that would internally contain a `grpc.Server`. It
```
s/`xds.Server`/`xds.GRPCServer`

https://github.com/grpc/grpc-go/blob/8736dcd05b735913226c0c8bf787db48a13a7129/xds/server.go#L86

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+Create an `xds.Server` struct that would internally contain a `grpc.Server`. It
+would inject its own `StreamServerInterceptor` and `UnaryServerInterceptor`s
+into the `grpc.Server`. The interceptors would be controlled by an
+`atomic.Value` that would update with the configuration. When
```
Could we skip this line about using an `atomic` to update the interceptors? It seems very implementation specific and could change anytime. Currently, we do have an interceptor which does nothing, and is not changed based on any configuration received. When we start supporting changing of the interceptors based on received configuration, we could use an `atomic` or a `mutex`.

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+TODO: It looks like there is a lifetime issue for the XdsClient, as RPCs can
+continue after the listener is closed and Serve() returns.
+
+Because service registration is done on a method in the generated code, and
```
This is no longer true. We have added a `ServiceRegistrar` interface to the `grpc` package which contains a single method `RegisterService`, and this is implemented by both `grpc.Server` and `xds.GRPCServer`. The protoc-gen-go has also been updated to emit code where the registration function accepts this interface instead of a concrete type. So, users should be able to directly pass `xds.GRPCServer` to these registration functions, and the underlying `grpc.Server` is no longer exported.

https://github.com/grpc/grpc-go/blob/8736dcd05b735913226c0c8bf787db48a13a7129/server.go#L558

easwars commented on A36-xds-for-servers.md:
````diff
@@ -0,0 +1,316 @@
+```
+
+
+To allow passing configuration, the `xds.Server` will be responsible for
````
This section needs updating as well. The current plan can be summarized as follows:

- An API to create a server-side `XdsCredentials` will be exposed. This will be similar to the [NewClientCredentials](https://github.com/grpc/grpc-go/blob/8736dcd05b735913226c0c8bf787db48a13a7129/credentials/xds/xds.go#L61), and would be called `NewServerCredentials`.
- The `xds.GRPCServer` will look through the `ServerOptions` list passed to it and will use a type assertion to determine whether or not the user has specified `XdsCredentials`.
- If the user has specified `XdsCredentials`, any information required by the credentials implementation will be made available by exposing a function on a wrapped `net.Conn` passed to the [ServerHandshake](https://github.com/grpc/grpc-go/blob/8736dcd05b735913226c0c8bf787db48a13a7129/credentials/xds/xds.go#L352) method.
- The `xds.GRPCServer` will be responsible for wrapping the raw `net.Conn` corresponding to the incoming connection before invoking the `ServerHandshake` method.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+## Abstract
+
+Bring xDS-based configuration to Servers. xDS has many features and this gRFC
+will not provide support for any of them in particular. However, it will be
```
s/it will be provide/it will provide/

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+configure the API, gRPC should log the errors at a default-visible log level. If
+the xds bootstrap is missing or invalid, implementations would ideally fail
+server startup, but a permanent hang is also acceptable. "Initial configuration"
+does not have to include server credentials.
```
Not clear what this means. Does it mean the initial config received from the control plane may not include TLS configuration but a later one may? Or by "server credentials" you mean the credentials needed to talk to the control plane? In this context "server" is ambiguous because it could mean the xDS control plane or the gRPC xDS-server under question.

markdroth reviewed (changes requested):
Thanks for writing this up, Eric!

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/proposal/214#-:-MOI_lrMEiAMq1E608tR:b-hl0msq)* status: all files reviewed, 20 unresolved discussions (waiting on @easwars and @ejona86)

___
*[A36-xds-for-servers.md, line 57 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOHpfYGDfLBw-d_B8tJ:-MOHpfYHBBuGymjOHMAt:buriu40) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L57)):*
> ```Markdown
> of the existing API.
> 
> Credential configuration will be managed separately by providing a
> ```

This paragraph seems like it might belong in the security gRFC instead of this one.  Or maybe leave it here but add something like "see gRFC A29 for details"?
___
*[A36-xds-for-servers.md, line 65 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOHx9CP4PIHiQWa9L7q:-MOHx9CP4PIHiQWa9L7r:b-75cvsv) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L65)):*
> ```Markdown
> 
> When starting the XdsServer, it will start communication with xDS to receive
> initial configuration. `bind()` and `listen()` can happen immediately, but
> ```

I think it's worth pointing out that calling `listen()` before the xDS resource is obtained is actually worse for clients, because even without calling `accept()`, it will still cause timeout delays before pick_first can fail over to the next address in the list.  So I think implementations should prefer to defer calling `listen()` until after the xDS resource is obtained.  That's what we're planning to do in C-core.

Also, the language here seems to imply that `bind()` should be done immediately even if `listen()` is delayed.  Is there actually any reason for that?  Based on what @yashykt and I are looking at in C-core, it seems like the only case where it would matter that `bind()` happens before we get the xDS resource is a server using an ephemeral port.  And that seems like such a niche case that I think I'd be okay saying that we just don't support it for the xDS-enabled server API.  (To be clear, it does seem better to call `bind()` earlier if possible, so that we can support ephemeral ports, but I think it's a nice-to-have, not a hard requirement.  In C-core, it looks like that may be hard to de-couple, so at least initially we will probably start by deferring both `bind()` and `listen()` until after we have the xDS resource.  We can revisit that later if someone actually wants to use ephemeral ports with xDS.)
___
*[A36-xds-for-servers.md, line 80 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOHoPfA0OI2pOXbO3Oz:-MOHoPfB-lUDEoFaS5nW:b6hdak2) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L80)):*
> ```Markdown
> ```json
> {
>   "grpc_server_resource_name_id": "example/resource",
> ```

Given our recent discussion about what we're going to want to do for new-style resource names, how about doing something like this:

```
  // A template for the name of the Listener resource to subscribe to
  // for a gRPC server.
  // The token "%s", if present in this string, will be replaced with
  // the IP and port on which the server is listening.
  "server_listener_resource_name_template": "grpc/server?xds.resource.listening_address=%s",
```

This seems more flexible for deployments, and it has the nice property that we'd avoid hard-coding the "xds.resource.listening_address" part.
___
*[A36-xds-for-servers.md, line 90 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIHFsF1VrKJ2rHfvO_:-MOIHFsF1VrKJ2rHfvOa:b-sfdw3u) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L90)):*
> ```Markdown
> `envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
> value of `grpc_server_resource_name_id` from the bootstrap,
> `?udpa.resource.listening_address=`, and the listening address to construct the
> ```

s/udpa/xds/
___
*[A36-xds-for-servers.md, line 93 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIHIsYDe4X1wUautD8:-MOIHIsYDe4X1wUautD9:b-sfdw3u) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L93)):*
> ```Markdown
> resource name. For example, with an address of `[::]:80` and the example
> bootstrap field above, the resource name would be
> `example/resource?udpa.resource.listening_address=[::]:80`. Note that the `[]`
> ```

s/udpa/xds/
___
*[A36-xds-for-servers.md, line 96 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOHo26w6p_V9H_BAa2C-r1-96:-MOIWDwn5AHT--q1oOlb:b-9athw6) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L96)):*
<details><summary><i>Previously, sanjaypujare wrote…</i></summary><blockquote>

There is no `listening_address` field. It is just encoded in the resource name in the request. So when you match the name in the response to the name in the request you are indirectly verifying the value of the `listening_address`. In that case it should not be difficult to match that part with the `[address][Listener.address]` value either.
</blockquote></details>

Hmm, good question.  Maybe the XdsClient can verify that the `Listener` resource has its `address` field set to the same address as in the resource name.  Although that might be difficult if we use the template approach I suggest above instead of hard-coding "xds.resource.listener_address=" in our implementation.
___
*[A36-xds-for-servers.md, line 120 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIXCa87o7bi-uUNyd6:-MOIXCa87o7bi-uUNyd7:bsgajl4) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L120)):*
> ```Markdown
>     EXTERNAL = 2;
>   }
>   google.protobuf.UInt32Value destination_port = 8; // Always fail match
> ```

Do we fail to match even if the destination port actually matches the port we're listening on?
___
*[A36-xds-for-servers.md, line 125 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIYV1P40mgCXusfkYw:-MOIYV1P40mgCXusfkYx:b-vuxh6t) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L125)):*
> ```Markdown
>   repeated core.v3.CidrRange source_prefix_ranges = 6;
>   repeated uint32 source_ports = 7;
>   repeated string server_names = 11; // May always fail match if TLS unsupported
> ```

The comment here is a little confusing.  What's the intended behavior here?  (It's not actually entirely clear to me what this field actually does in Envoy.)
___
*[A36-xds-for-servers.md, line 126 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIXyItDRLfpzFqplcw:-MOIXyItDRLfpzFqplcx:b-lu5twz) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L126)):*
> ```Markdown
>   repeated uint32 source_ports = 7;
>   repeated string server_names = 11; // May always fail match if TLS unsupported
>   string transport_protocol = 9; // Always fail match
> ```

Why would this always fail?  It looks like the supported values are "raw_buffer" and "tls", to select based on whether the connection is encrypted or not.
___
*[A36-xds-for-servers.md, line 127 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIYJFM1NsQg6hcJpqI:-MOIYJFM1NsQg6hcJpqJ:bmkxxz4) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L127)):*
> ```Markdown
>   repeated string server_names = 11; // May always fail match if TLS unsupported
>   string transport_protocol = 9; // Always fail match
>   repeated string application_protocols = 10; // Always fail match
> ```

Always fail even if this says "h2"?
___
*[A36-xds-for-servers.md, line 131 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOI_OqF9RyAjMOsbUbR:-MOI_OqG8a0yvuyD5mY3:b-5uzwa3) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L131)):*
> ```Markdown
> ```
> 
> All fields are "supported," however, we know that some features are unsupported.
> ```

I don't quite understand the meaning of this paragraph.

Should we reject a config that specifies filters, or do we just ignore the filters for now?  If we just ignore them for now, then what happens when we later add support for filters?  Will we be in a situation where a config that did previously work will then be NACKed?

In particular, I'm wondering about the HTTP connection manager (HCM) filter, which should always be the last filter in the chain.  Should we verify that that's there, since we know that we're going to need some functionality from the HCM config later?
___
*[A36-xds-for-servers.md, line 141 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIZrxH6UzCYmoyIQ1d:-MOIZrxH6UzCYmoyIQ1e:b45ga6z) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L141)):*
> ```Markdown
> no longer in use; they are a resource and must not be leaked.
> 
> This overall "wrapping" server API has many language-specific ramifications. We
> ```

I wonder if we should actually split the individual language sections out into their own gRFCs with `L` numbers?  We're ultimately going to need a separate section for every wrapped language.
___
*[A36-xds-for-servers.md, line 284 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIZK4PDmc1ef5W9Y0n:-MOIZK4PDmc1ef5W9Y0o:bwty7jn) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L284)):*
> ```Markdown
> impact is hoped to be low as servers will generally receive their initial
> configuration quickly (on the order of a second) and most clients will likely be
> using a variant of round-robin. We'd also encourge the xDS control plane to
> ```

Why do we encourage the xDS server to avoid handing out addresses of not-yet-started servers?  I don't see the rationale for that.


<!-- Sent from Reviewable.io -->

yashykt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/grpc/proposal/214)* status: 0 of 1 files reviewed, 21 unresolved discussions (waiting on @easwars, @ejona86, and @markdroth)

___
*[A36-xds-for-servers.md, line 101 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIyjSbDn5DmoDcwwam:-MOIyjSbDn5DmoDcwwan:bcoj6oj) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L101)):*
> ```Markdown
> the resource if the address does not match.
> 
> Although `Filter`s will not be observed initially, the `FilterChain` contains
> ```

Is this paragraph meant to be a placeholder for future changes? i.e. do we expect the types that we use to change in the future?
___
*[A36-xds-for-servers.md, line 103 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOIvzpm5PRYncKh92cC:-MOIvzpm5PRYncKh92cD:b-jvc5t1) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L103)):*
> ```Markdown
> Although `Filter`s will not be observed initially, the `FilterChain` contains
> data that may be used. When looking for a FilterChain, the standard matching
> logic must be used. Each `filter_chain_match` of the repeated
> ```

FYI currently, implementation-wise C Core and Go are accepting only a single `filter_chain` entry, so there is no matching logic in place at the moment. We aren't looking at `default_filter_chain` either. The only check that we are expecting to perform from the single entry's `filter_chain_match` is to check whether the application_protocol mentions "h2".


<!-- Sent from Reviewable.io -->

markdroth reviewed (changes requested):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/grpc/proposal/214)* status: all files reviewed, 21 unresolved discussions (waiting on @easwars and @ejona86)

___
*[A36-xds-for-servers.md, line 80 at r1](https://reviewable.io/reviews/grpc/proposal/214#-MOHoPfA0OI2pOXbO3Oz:-MOJ0FEY7cj9XcgufNiv:brooi1f) ([raw file](https://github.com/grpc/proposal/blob/b825f33e5dbb491b048ca6a97491456dca294a8b/A36-xds-for-servers.md#L80)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Given our recent discussion about what we're going to want to do for new-style resource names, how about doing something like this:

```
  // A template for the name of the Listener resource to subscribe to
  // for a gRPC server.
  // The token "%s", if present in this string, will be replaced with
  // the IP and port on which the server is listening.
  "server_listener_resource_name_template": "grpc/server?xds.resource.listening_address=%s",
```

This seems more flexible for deployments, and it has the nice property that we'd avoid hard-coding the "xds.resource.listening_address" part.
</blockquote></details>

Although actually, if we do use the template here, then the parameter name should probably not start with "xds.", since that namespace is reserved for parameters added by the client implementation, not those added by the user.


<!-- Sent from Reviewable.io -->

ejona86 commented:
Responses to @markdroth

> So I think implementations should prefer to defer calling listen() until after the xDS resource is obtained.

The very next sentence stated "`listen()` should be delayed if possible". The rational section describes the issues listening causes.

Given our other discission, I think this may change to "either don't call listen(), or call listen() and do accept()+close()".

> Also, the language here seems to imply that bind() should be done immediately even if listen() is delayed.

It sayed "can". Changed to "may." Java does not bind before receiving the xDS Listener, so that definitely was not the intention. And yes, ephemeral ports won't have a port chosen at that point.

> Given our recent discussion about what we're going to want to do for new-style resource names, how about doing something like this: (server_listener_resource_name_template snippet)

Yes, we should change it. Although I'll note that this will break experimental Java users today. But I think we can just add a new experimental flag to the bootstrap and leave the current flag's behavior unchanged. When we remove the grpc_server_resource_name_id flag that's when those clients will break, but they should have migrated to a new experimental flag or the default-on behavior once it is available.

> Although actually, if we do use the template here, then the parameter name should probably not start with "xds.", since that namespace is reserved for parameters added by the client implementation, not those added by the user.

It isn't a new-style name so it was already wild-west. But, yeah, good point. I guess with this templating thing we don't need a well-defined name and can just let every server define their own, since now it _isn't_ a protocol between the client and server as it was when this was implicit. CC @sanjaypujare

>> `envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
value of `grpc_server_resource_name_id` from the bootstrap,
`?udpa.resource.listening_address=`, and the listening address to construct the

> s/udpa/xds/

Looks like Java, and thus probably TD, already used 'udpa'. That can be changed, but "we had all agreed" already to the old name. We can either leave it as-is for the old-style name and improve things for new-style names or change it now even though "it was already decided and agreed on and implemented." I understand how this happened though; this is due to a more clear future direction since we fleshed out a federation design. CC @sanjaypujare

>>   google.protobuf.UInt32Value destination_port = 8; // Always fail match

> Do we fail to match even if the destination port actually matches the port we're listening on?

Yes. The Listener's `address` is what is important for the local port. `destination_port` is only applicable for `use_original_dst`. It's documentation: "Optional destination port to consider when use_original_dst is set on the listener in determining a filter chain match."

>>   repeated string server_names = 11; // May always fail match if TLS unsupported

> The comment here is a little confusing. What's the intended behavior here? (It's not actually entirely clear to me what this field actually does in Envoy.)

Changed to "Always fail match for non-TLS connections". I think I had intended to let this to have "always fail" logic before XdsServerCredentias is implemented, but I think that actually doesn't help much. This uses SNI to determine what host the client thinks it is connecting to.

>>   string transport_protocol = 9; // Always fail match

> Why would this always fail? It looks like the supported values are "raw_buffer" and "tls", to select based on whether the connection is encrypted or not.

It changes based on envoy.filters.listener.tls_inspector; aka connection sniffing. I seriously considered only matching the value "raw_buffer", but the entire thing makes me uneasy as it is really easy to accidentally use wrong.


>>   repeated string application_protocols = 10; // Always fail match

> Always fail even if this says "h2"?

Yep. This is more tls_inspector goo; it is only set by that filter (at least the docs say so).


>> All fields are "supported," however, we know that some features are unsupported.

> I don't quite understand the meaning of this paragraph.
>
> Should we reject a config that specifies filters, or do we just ignore the filters for now?

Filters will be ignored initially, just like on client-side. That was at defined at "Although `Filter`s will not be observed initially..."

But this discussion of "fields" only applies to FilterChainMatch.

You, Doug, and me had discussed checking filters and decided to just mirror client-side. Any issues that arise will be no worse than we will already suffer.

>> impact is hoped to be low as servers will generally receive their initial
configuration quickly (on the order of a second) and most clients will likely be
using a variant of round-robin. We'd also encourge the xDS control plane to

> Why do we encourage the xDS server to avoid handing out addresses of not-yet-started servers? I don't see the rationale for that.

In the context of EDS, it is best not to send clients to unhealthy/not ready backends.

ejona86 commented:
Responses to @yashykt
>> Although `Filter`s will not be observed initially, the `FilterChain` contains

> Is this paragraph meant to be a placeholder for future changes? i.e. do we expect the types that we use to change in the future?

At some point in the future we expect to support filters. But for this gRFC it will not change; adding filter support will be some later effort.

> FYI currently, implementation-wise C Core and Go are accepting only a single filter_chain entry, so there is no matching logic in place at the moment. We aren't looking at default_filter_chain either. The only check that we are expecting to perform from the single entry's filter_chain_match is to check whether the application_protocol mentions "h2".

That's fine since that code is not enabled by default. But every part of that behavior would need to be changed/enhanced before GA.

ejona86 commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,316 @@
+  ConnectionSourceType source_type = 12;
+  repeated core.v3.CidrRange source_prefix_ranges = 6;
+  repeated uint32 source_ports = 7;
+  repeated string server_names = 11; // May always fail match if TLS unsupported
```
Doh. This appears to rely on tls_inspector. https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/sni . The proto documentation wasn't quite as clear: "Those values will be compared against the server names of a new connection, when detected by one of the listener filters."

transport_protocol is only useful for sniffing, which we aren't interested in supporting. application_protocols could potentially be useful in the future, but not so much now (always h2). But server_names is more useful... Need to decide if/how we handle it.

sanjaypujare replied:
I vote against supporting this since we don't have tls_inspector kind of functionality. It's not worth the trouble to add this support esp considering there might not be use-cases.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+server startup, but a permanent hang is also acceptable.
+
+The XdsServer does not have to wait until server credentials (e.g., TLS certs)
+are available before accepting connections; since XdsServerCredentials might not
```
The wording is a bit confusing. Whether using or not `XdsServerCredentials` is one thing. And the XdsServer not having to wait for the certs to be available before accepting connections even when `XdsServerCredentials` is used is a different thing. I think what is being specified here is the case when `XdsServerCredentials` is used. The server is free to start accepting connections even if the corresponding TLS certs are not available yet.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+address, like `::` or `0.0.0.0`) to a new XdsClient API to start a watch for a
+`envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
+value of `grpc_server_resource_name_id` from the bootstrap,
+`?udpa.resource.listening_address=`, and the listening address to construct the
```
I think there is talk of changing this to `?xds.resource.listening_address=` to preserve the xDS nomenclature. Pls check

ejona86 replied:
Mark mentioned that as well. However, with other changes the string is now determined by the bootstrap file. I left it in as an example and changes s/udpa/xds/, but it is really up to the bootstrap file.

sanjaypujare replied:
Yes, I saw Mark's comments later. He made his comments using the "Reviewable" stuff and I don't know how to integrate with that stuff (without using Reviewable myself). Also those comments don't show up on in the "Files changed" tab but only in the "Conversation" tab :-(

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+    SAME_IP_OR_LOOPBACK = 1;
+    EXTERNAL = 2;
+  }
+  google.protobuf.UInt32Value destination_port = 8; // Always fail match
```
"Always fail match" means "fail the match if the field is present"? Just trying to clarify

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+    EXTERNAL = 2;
+  }
+  google.protobuf.UInt32Value destination_port = 8; // Always fail match
+  repeated core.v3.CidrRange prefix_ranges = 3;
```
If `prefix_ranges` is to be matched why is [`address_suffix`](https://github.com/envoyproxy/envoy/blob/928a62b7a12c4d87ce215a7c4ebd376f69c2e080/api/envoy/config/listener/v3/listener_components.proto#L113) left out?

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+
+Although `Filter`s will not be observed initially, the `FilterChain` contains
+data that may be used. When looking for a FilterChain, the standard matching
+logic must be used. Each `filter_chain_match` of the repeated
```
Filter and related objects (FilterChain and FIlterChainMatch) are very Envoy specific. For gRPC we have already modified the listener query to supply the resource name (`example/resource?udpa.resource.listening_address=[::]:80` in the above example) which Envoy doesn't use. We can also modify the response specification to say that the response should not contain any filterChainMatch since they don't apply that well to gRPC. The rationale is that the control plane will have to make changes to support gRPC specific listener query so they will also make changes to send us responses without any filter data.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+  ConnectionSourceType source_type = 12;
+  repeated core.v3.CidrRange source_prefix_ranges = 6;
+  repeated uint32 source_ports = 7;
+  repeated string server_names = 11; // Always fail match for non-TLS connections
```
And in case of TLS? This is meant to be for SNI and it will be very difficult to use this for SNI in most of our gRPC stacks: we'll need to make these filterChains available to the TLS layer in gRPC which can tell us the server-name used in the TLS handshake and we have to use that to then match the filterChain which will give us the DownstreamTlsContext that we then use to retrieve the certs that we need to pass back into the TLS layer to continue the handshake. Also SNI is used more in shared web infrastructure so I don't know how common it is in gRPC and service mesh scenarios.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+  repeated core.v3.CidrRange source_prefix_ranges = 6;
+  repeated uint32 source_ports = 7;
+  repeated string server_names = 11; // Always fail match for non-TLS connections
+  string transport_protocol = 9; // Always fail match
```
Why not accept "tls" ?

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+  repeated uint32 source_ports = 7;
+  repeated string server_names = 11; // Always fail match for non-TLS connections
+  string transport_protocol = 9; // Always fail match
+  repeated string application_protocols = 10; // Always fail match
```
I think we should match when it contains "h2"

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,324 @@
+would also be difficult to share the XdsClient configuration between the
+interceptor and credentials. It simply cannot be used in its current state.
+
+We chose the XdsServer approach over creating a new server-side plugin API. In
```
Even with the `XdsServerBuilder` approach, we can offer a convenience function for ease of use. For example in  Java:
```
ServerBuilder<?> serverBuilder = Grpc.newServerBuilderForPort(host, port, serverCreds);
```
host is the new argument and if you pass something like `"xds:///localhost"` then the XdsServerBuilder is returned. Similar to how you engage xDS on the client side. Of course you will then have to cast the ServerBuilder to XdsServerBuilder if you want to use other functionality of the builder.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -54,57 +54,71 @@ also allow it access to the server to change behavior (e.g., via an
 When starting the XdsServer, it will start communication with xDS to receive
-initial configuration. `bind()` and `listen()` can happen immediately, but
-`accept()`ing on the port will be delayed until the initial configuration is
+initial configuration. `bind()` and `listen()` may happen immediately, but
```
I think we should actually write down all of the considerations we've discussed about this, so that future readers understand the implications.  Here's some possible wording; feel free to change as needed:

"""
The server should ideally avoid calling `listen()` on the port until after it has obtained the initial configuration, so that clients attempting to connect to the server will quickly fail, thus preserving reasonable pick_first behavior.  If for some reason that is not feasible, a reasonable alternative would be to start listening but immediately `accept()` and then `close()` each incoming connection attempt, since this will result in essentially the same behavior on the client.  A less desirable alternative would be to call `listen()` and then simply not call `accept()` on incoming connection attempts, since that may cause the connection attempts to timeout instead of immediately failing, depending on the size of the accept queue.

The server must call `bind()` before calling `listen()`, but some implementations cannot make those two calls at different times.  However, if the implementation desires to support ephemeral ports, it must call `bind()` to determine the actual port being used before it can request the xDS resource.  Therefore, server implementations that cannot call `bind()` and `listen()` separately must choose between (a) not supporting ephemeral ports and calling `bind()` and `listen()` after receiving the xDS resource and (b) supporting ephemeral ports by calling `bind()` and `listen()` before receiving the xDS resource, using the `accept()` and `close()` approach.  (Note that we do not expect that ephemeral ports are likely to be a common case with xDS in the first place.)

Note that the above considerations about when to call `bind()` and `listen()` arise not only when the server first starts up but also when the server receives an xDS update indicating that the required resource has been removed.  Although the same considerations apply to both cases, implementations may choose different trade-offs in each case.  For example, an implementation may choose at startup to defer calling `listen()` until it receives the xDS resource, but when it receives an update telling it that the xDS resource has been removed, it may be easier to `accept()` and `close()` incoming connections rather than to stop listening.
"""

I realize that some of this is currently covered in the "Rationale" section, but I think it's a little easier to understand if all of the considerations are described in one place.  I think the "Rationale" section can basically just point out that an alternative way of ameliorating the client pick_first behavior would be to implement Happy Eyeballs on the client side; it won't need to cover anything else, because the rest will be covered here.

markdroth replied:
As per discussion in https://github.com/grpc/grpc/pull/24956#discussion_r541417034, we should also document the fact that when the Listener resource is removed by the xDS server, we will also close any pre-existing connections.

CC @yashykt

markdroth reviewed (commented):
This looks really good!

Just a couple of comments remaining.  One of them is mostly cosmetic (being a bit clearer in terms of how we describe the constraints around `bind()` and `listen()`).  The other is actually a content change, unless we can think of a better way to do the address check.

Please let me know if you have any questions.  Thanks!

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+* Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2020-11-19
+* Discussion at: <google group thread> (filled after thread exists)
```
Did you create the thread? You should do so ASAP if not since it needs to wait at least 2 weeks.

ejona86 replied:
Well, given that Go's section is apparently all wrong, I think we'd need to at least get it in working shape or remove it before then to avoid confusion.

easwars replied:
I pushed a commit which updates the Go section. PTAL @dfawley @ejona86

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+the central plumbing and APIs to add support for specific features and have them
+work without additional user code changes.
+
+There's multiple pieces involved:
```
English nit: "There are"

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+## Proposal
+
+Each language will create an "XdsServer" API that will wrap the normal "Server"
+API. This will allow it to see the listening ports, start event, stop event, and
+also allow it access to the server to change behavior (e.g., via an
+interceptor). Users will use the XdsServer API to construct the server instead
```
I'm not sure this is true for Go, at least.  We don't have hooks for start/stop, and the current xds server API doesn't allow injecting interceptors or direct access to the underlying grpc server.

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+
+When starting the XdsServer, it will start communication with xDS to receive
+initial configuration. `bind()` and `listen()` may happen immediately, but
+`accept()`ing on the port must be delayed until the initial configuration is
+received. `listen()` should be delayed if possible, but commonly needs to happen
```
Do we still want to delay `Accept` before the initial configuration is received, or do the `Accept`+`Close` approach that we decided if we receive a bad config?  cc @markdroth

markdroth replied:
See my comment from yesterday.  I suggested a fairly long explanation of all of the constraints around when to call `bind()`, `listen()`, `accept()`, etc.

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+initial configuration. `bind()` and `listen()` may happen immediately, but
+`accept()`ing on the port must be delayed until the initial configuration is
+received. `listen()` should be delayed if possible, but commonly needs to happen
+immediately after `bind()` for implementation-specific reasons. If initial
+configuration fails, the server will continue delaying and server startup will
+stall until it succeeds. Each language will provide an XdsServer-specific API to
```
I believe this should be updated to mention `Accept`+`Close` in the event of initial configuration failure - or if the resource is subsequently removed.

dfawley commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,332 @@
+received. `listen()` should be delayed if possible, but commonly needs to happen
+immediately after `bind()` for implementation-specific reasons. If initial
+configuration fails, the server will continue delaying and server startup will
+stall until it succeeds. Each language will provide an XdsServer-specific API to
+inform the application of initial communication failures. If the user does not
```
Does Java/C have (or plan to have) an API to expose this?  I don't think we designed anything like this for Go (yet?).  @easwars @yashykt

yashykt replied:
C does not have one yet, but it is in the works.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -104,13 +104,15 @@ is _not_ percent encoded, as this string is not a URI.
-Although `Filter`s will not be observed initially, the `FilterChain` contains
-data that may be used. When looking for a FilterChain, the standard matching
-logic must be used. Each `filter_chain_match` of the repeated
+Although `FilterChain.filters`s will not be observed initially, the
```
Not clear about "observed initially". Are you saying "initially" as in the first few updates from the control plane may not contain `FilterChain.filters` but will be present later? Or "initially" as in the first/initial gRPC implementation might choose to ignore `FilterChain.filters`.

ejona86 replied:
Reworded to say "as part of this gRFC"

ejona86 commented:
Okay, I've worked through the comments. I still need a design for Go, @dfawley @easwars; if we don't have one soon I'll need to rip out Go to move forward. It seems like you mostly have one, but I just don't know what it is. If you want to make direct edits to the Go section (and I will review them), that doesn't bother me. I'd like to send out notice to the mailing list in the next few days, but I don't want the Go section to be misleading when I do so.

The structure has changed dramatically now, mostly around the listen() vs accept()+close() stuff, but it was far-reaching. You probably want to take a look at "Explicitly define how to shut down existing connections", as we had not previously discussed interactions with the GOAWAY race. I don't expect my proposal to be contentious, you just may need to consider it when designing/implementing.

ejona86 commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,380 @@
+### Related Proposals:
+
+ * [A27: xDS-Based Global Load Balancing][A27]
+ * [A29: xDS-Based Security for gRPC Clients and Servers][A29]
```
Yes, because it is still in-review. No, I don't want to reference a PR, as I'm unlikely to ever notice it needs fixing.

ejona86 commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,385 @@
+### Go
+
+Create an `xds.GRPCServer` struct that would internally contain an unexported
+`grpc.Server`. It would inject its own `StreamServerInterceptor` and
```
@easwars, if a user has their own interceptor, how will this interceptor call their interceptor?

ejona86 commented on A36-xds-for-servers.md:
````diff
@@ -0,0 +1,385 @@
+```
+
+
+Package `credentials/xds` exports a `NewServerCredentials()` function which
````
@easwars, will the `NewServerCredentials()` have an argument for 'fallback' credentials if the xds server does not provide credentials?

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -95,23 +136,27 @@ The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
 
 XdsServer will use the normal `XdsClient` to communicate with the xDS server.
 There is no default value for `server_listener_resource_name_template` so if it
```
Didn't we say that it was going to default to `%s`?

markdroth replied:
Okay.  This means that users basically need to add this to their bootstrap files before they can use xDS with a gRPC server, but I guess that's okay.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -95,23 +136,27 @@ The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
-Although `FilterChain.filters`s will not be observed initially, the
+is not present in the bootstrap then server creation or start will fail or the
+XdsServer will become "not serving". The XdsServer will pass the listening
+address (commonly using a wildcard IP address, like `::` or `0.0.0.0`) to a new
```
I think the template expansion will actually be done by the XdsServer code, not by the XdsClient.  The XdsServer will expand the template and then use the expanded name to start a Listener watch on the XdsClient.  The Listener watch will by done via the existing XdsClient API for Listener watches (i.e., the same one used in the gRPC client).

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -95,23 +136,27 @@ The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+To be useful, the xDS-returned Listener must have an
+[`address`][Listener.address] that matches the listening address provided. The
+Listener's `address` would be a TCP `SocketAddress` with matching `address` and
+`port_value`. The XdsClient must be "not serving" if the address does not match.
```
s/XdsClient/XdsServer/

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -148,22 +193,28 @@ result that can be hard-coded. This applies to `destination_port` which relies
+
+The overall "wrapping" server API has many language-specific ramifications. We
+show each individual language's approach. For C++ and wrapped languages we just
+show a rough sketch of how they would be done and expect a followup gRFC for
```
We'll actually probably just backfill the details into this gRFC, rather than creating a separate one.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -325,8 +383,18 @@ ServerCredential to use, which is unclear how that would be accomplished.
+
+* C++/wrapped languages. The surface APIs should be relatively easy compared to
+  the C core changes. The C core changes are being investigated by @markdroth
+  and the plan is to cover them in a separate gRFC. So this gRFC just tries to
```
As mentioned above, we'll backfill the details into this gRFC in a future PR.

markdroth reviewed (commented):
I think this is getting close to being ready!  My comments are mostly minor at this point.

ejona86 reviewed (commented):
I agree that "unavailable" needs to be tweaked. But I don't want to fully-define it at that location. I'll figure it out soon; the rest should be fixed now.

ejona86 commented on A36-xds-for-servers.md:
```diff
@@ -63,17 +64,41 @@ ServerCredentials types to XdsServer and they will be honored. Note that RBAC is
+
+To serve RPCs, the XdsServer must have its xDS configuration, provided via a
+Listener resource and potentialy other related resources. When its xDS
+configuration is unavailable the server must be in a "not serving" mode. This is
```
I replaced "unavailable" with "missing or known". I think that addresses the issue and is terse.

yashykt commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+The xDS client must NACK the Listener resource if `Listener.listener_filters` is
+non-empty.
+
+### FilterChainMatch
```
Let's also specify the behavior of the xDS client when there are no valid filter chains present.

yashykt commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+must be used. The [most specific `filter_chain_match`][matching logic] of the
+repeated [`filter_chains`][Listener.filter_chains] should be found for the
+specific connection and if one is not found, the `default_filter_chain` must be
+used. The following is a snippet of [all current fields of
```
If no `default_filter_chain` is present, the connection is closed.

easwars replied:
Was that a question or a statement? If that's a question, I'd like to know the answer for the same.

yashykt replied:
It's a statement, potentially serving as the start of a discussion if we want to do something else when there are are no filter chains that we can use.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once; `build()` should throw if called a
+second time. To allow the `XdsServerBuilder` to create addition servers if
```
s/addition/additional/

But I don't understand this. If subsequent calls to `build()` should throw then how do you propose allowing creation of additional servers with the same builder? Or is that meant to be a future enhancement?

ejona86 replied:
Maybe `XdsServer` there would be more clear. I can also drop a hint that this relates to "serving"/"not serving." The point is for the implementation to be able to construct more servers for internal use, not to pass more servers to the user.

sanjaypujare replied:
Okay, got it. The "not serving" to "serving" transition will requires XdsServer to rebuild a new server to be able to serve. Not sure if the internal delegate NettyServerBuilder allows more calls to build.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+is restored. However, XdsServer must accept configuration changes provided by
+the xDS server, including resource deletions. If that causes the XdsServer to
+lack essential configuration (e.g., the Listener was deleted) the server would
+need to enter "not serving" mode. "Not serving" requires existing connections to
```
How about the "Not serving" to "Serving" transition? Specifically:

- this transition will automatically start processing RPCs without any intervention by the application, correct? i.e. it is as though `start()` was called on the XdsServer object (but without the application having to call it)
- this will require recreating the internal/delegate Server object since the delegate does not support restarting after shutting down. Is that correct?

It will be good to add some clarification for this.

ejona86 replied:
> this transition will automatically start processing RPCs without any intervention by the application, correct? i.e. it is as though start() was called on the XdsServer object (but without the application having to call it)

Yes. The application had already called `start()` as is expecting the server to be running. It was the library that "called stop()" so it is its responsibility to resume and restore the expected application state when able.

> this will require recreating the internal/delegate Server object since the delegate does not support restarting after shutting down. Is that correct?

Correct. This will vary with the language-specific API and the approach taken to become "not serving." But assuming XdsServer calls `server.stop()` on the wrapped server (the internal one), then it'd need to create a new one to replace it when calling `start()`.

I didn't go into more detail _here_ because that is language and implementation-specific. I did mention in the Java section that we may need to rebuild the server, and that was to allow for things here. But it isn't required to re-create the server; I left open the option for you to use a ProtocolNegotiator to immediately close() connections. Such an approach wouldn't have needed to re-create the server.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+
+If not specified, a default implementation of `XdsServingStatusListener` will be
+used. It will log the exception and log calls to `onServing()` following a call
+to `onNotServing()` at WARNING level.
```
Shouldn't the first `onServing()` call also log i.e. the first time the server goes into serving mode?

ejona86 replied:
No. The point was to only log when something is going wrong, or has been resolved such that that previous log is no longer correct. We still don't want to log during normal operation.

ejona86 commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+XdsServer will become "not serving". XdsServer will perform the `%s` replacement
+in the template (if the token is present) to produce a listener name. The
+XdsServer will start an XdsClient watch on the listener name for a
+`envoy.config.listener.v3.Listener` resource. No special character handling of
```
Add a note that v2 would not be supported.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -68,24 +68,25 @@ XdsServerCredentials.
-would naturally attempt connections to any remaining addresses to quickly find
-an operational backend. However, the `accept()`+`listen()` approach will be
-improperly detected as server liveness for TCP heath checking.
+configuration is missing or unknown the server must be in a "not serving" mode.
```
I think this still isn't very clear to readers who don't already know what we mean.  How about saying "does not exist or has not yet been received"?

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -197,8 +198,7 @@ on `use_original_dst`. It also applies to `server_names`, `transport_protocol`,
 show each individual language's approach. For C++ and wrapped languages we just
-show a rough sketch of how they would be done and expect a followup gRFC for
-more implementation details.
+show a rough sketch of how they would be done.
```
@yashykt, do we have enough details for the C++ API at this point that we can include them here?

yashykt replied:
Yeah, I think the C++ API is decided. We can add that here or create a separate gRFC. I'm fine with either of the two.

markdroth replied:
Let's include it here.

ejona86 replied:
This PR is pretty close, so let's do the C++ as a followup PR. It is just as easy, if not easier (you don't have to push to my branch), that way.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+
+### FilterChainMatch
+
+Although `FilterChain.filters`s will not be used as part of this gRFC, the
```
As we've discussed, we'll need to update this to reflect the following:
- We *will* actually look at the L4 filters, and we expect to see only the HCM there.
- As per gRFC A39, we will look at HTTP filters.  At the moment, the only HTTP filter supported on the server side will be the router filter.
- We need to define exactly how we're going to use the RouteConfiguration, especially with regard to possibly needing something like envoyproxy/envoy#13366.  (If we are using the RouteConfiguration, this also means that we may need to use RDS, and I assume that getting the RDS response would be required before the XdsServer can treat the config as being valid.)

markdroth reviewed (commented):
This looks really good!  I think the only things we need are the changes we've discussed about L4 filters, HTTP filters, and the RouteConfiguration.

Thanks again for driving this!

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+Listener's `address` would be a TCP `SocketAddress` with matching `address` and
+`port_value`. The XdsServer must be "not serving" if the address does not match.
+
+The xDS client must NACK the Listener resource if `Listener.listener_filters` is
```
Should we also NACK the `Listener` resource if `Listener.use_original_dst` is set?

ejona86 replied:
We don't care all that much, because it also requires iptables to have any effect and there's lots of unsupported fields we don't validate. But essentially we tossed a coin and decided, yeah, we'll validate it. Thanks for bringing it up; we hadn't considered it.

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+message FilterChainMatch {
+  enum ConnectionSourceType {
+    ANY = 0;
+    SAME_IP_OR_LOOPBACK = 1;
```
For the `SAME_IP_OR_LOOPBACK`, one can easily verify if the source IP of the incoming connection is from a loopback address. But do we also have to support the case where the host supports multiple addresses and we need to make sure that the source IP of the incoming connection is one of them? Also, does anything change based on whether the server is bound to an actual local address or `any` address?

ejona86 replied:
No, SAME_IP_OR_LOOPBACK means the source IP and dest IP are the same (or localhost); it doesn't support cases where a machine has multiple IPs and the IP selection is different between source and dest.

Nothing changes by the wildcard. It should be straight-forward to implement.

sanjaypujare commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+automatically (e.g., retry every minute).
+
+Communication failures do not impact the XdsServer's xDS configuration; the
+XdsServer should continue using the most recent configuration until connectivity
```
Is that true also for the very first time config is needed during `start()'? What's the expected behavior when the listener resource is missing at startup time? Should the server's `start()` (in Java) fail with the server going into permanent failure state?

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+To be useful, the xDS-returned Listener must have an
+[`address`][Listener.address] that matches the listening address provided. The
+Listener's `address` would be a TCP `SocketAddress` with matching `address` and
+`port_value`. The XdsServer must be "not serving" if the address does not match.
```
This might be more of an implementation related question. In Go, the layer in the `xdsClient` which handles the processing and parsing of the raw message received from the control plane (and makes the decision whether to ACK or NACK), is pretty low and does not have an idea about things like the resource name template configured in the bootstrap file. Currently this layer extracts the listening_address encoded as `ip:port` in the resource name and compares it against the socket address in the `address`. If they don't match, it rejects the resource. And once the resource is rejected here (and NACK'ed), upper layers (the ones which register watches on these resources) like the `xdsServer` will not even be notified about this. This means that an `xdsServer` will continue to remain blocked  waiting for the first resource or will continue using an earlier received good listener resource.

The question here is whether we can NACK the update when the listening address doesn't match and stop there or do we *have* to move the `xdsServer` to "not serving" state?
- This is not much of a concern if this issue happens as part of the first update since the server is anyways blocked on receiving the first good update and can continue to do so.
- But when the server is already using a good configuration and gets an update where the address doesn't match, it does raise an important question.

From the section on serving vs not serving, it does seem to be clear that we need to move to not-serving 
```
However, XdsServer must accept configuration changes provided by
the xDS server, including resource deletions. If that causes the XdsServer to
lack essential configuration (e.g., the Listener was deleted) the server would
need to enter "not serving" mode ...
```

How is Java/C++ handling this? @sanjaypujare @yashykt 

On an orthogonal note, the changes that we plan to make for CSDS which will ensure that all resources are parsed (even if some error out), and that upper layers are notified might help simplify things here.

ejona86 replied:
> If they don't match, it rejects the resource

No, it does not reject (NACK) the resource. The XdsClient validation is context-independent, so it won't be able to validate the listening_address is for the address the XdsServer needs. That was something we lost when we swapped away from the "custom xds.resource.listening_address protocol." If there is an address mismatch, XdsClient will ACK the resource but XdsServer will go "not serving." Not a great situation to be in, but it isn't the only time that sort of thing can happen and there's not really any alternatives.

I agree that _if we NACKed_, then XdsServer should just remain in whatever state it was previously (which would be easy). But we don't NACK.

easwars commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,401 @@
+There is no default value for `server_listener_resource_name_template` so if it
+is not present in the bootstrap then server creation or start will fail or the
+XdsServer will become "not serving". XdsServer will perform the `%s` replacement
+in the template (if the token is present) to produce a listener name. The
```
When we say `if the token is present` (which basically means this template is free-form), how are we going to ensure that gRPC can extract the actual listening_address out of it when we have to match it against the one stored in the `address` field of the received `Listener` resource?
@sanjaypujare @yashykt

ejona86 replied:
We don't extract the listening_address out of it. The matching of the address field is handled by the high-level XdsServer, not low-level XdsClient. XdsServer knows which address it is listening on.

easwars replied:
Thanks. So, the XdsServer replaces the token to generate the resource_name when requesting the Listener resource. But when matching the address field on the received Listener, it uses the address that it knows for sure that it is listening on. That makes sense.

ejona86 commented:
I pushed an important filter handling addition; I'd mentioned to some of you that was coming. It is _yet something else_ to take care of, but I hope isn't that bad because some client-side work that's already been happening should take care of it.

I made a smattering of fixes for nits. I'm aware some comments still need to be addressed/replied.

ejona86 commented:
If you see my "Make explicit that xds v3 is optional" commit message. Rest assured, it should have said _v2_. It happens that v3 isn't optional yet :-)

ejona86 reviewed (commented):
I think everything except the question of validating `use_original_dst` is resolved. `use_original_dst` will be a trivial change if we choose to make it, so I think this is very close and we should be looking at the final-ish round of reviews.

markdroth commented on A36-xds-for-servers.md:
```diff
@@ -0,0 +1,435 @@
+* Approver: markdroth
+* Status: In Review {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2020-11-19
```
Please update the date.

markdroth reviewed (approved):
This looks fantastic!  Thanks for nailing down all of these details!
